File: rust/src/benchmarks/spectralnorm.rs
use super::super::{Benchmark, helper};
use crate::config_i64;

pub struct Spectralnorm {
    size_val: i64,
    u: Vec<f64>,
    v: Vec<f64>,
    result_val: u32,
}

impl Spectralnorm {
    pub fn new() -> Self {
        let size_val = config_i64("Spectralnorm", "size");

        Self {
            size_val,
            u: vec![1.0; size_val as usize],
            v: vec![1.0; size_val as usize],
            result_val: 0,
        }
    }

    fn eval_a(&self, i: usize, j: usize) -> f64 {

        let ij = (i + j) as f64;
        1.0 / (ij * (ij + 1.0) / 2.0 + i as f64 + 1.0)
    }

    fn eval_a_times_u(&self, u: &[f64]) -> Vec<f64> {
        let n = u.len();
        let mut result = vec![0.0; n];

        let u_slice = u;

        for (i, r) in result.iter_mut().enumerate() {
            *r = u_slice.iter()
                .enumerate()
                .map(|(j, &u)| self.eval_a(i, j) * u)
                .sum();
        }

        result
    }

    fn eval_at_times_u(&self, u: &[f64]) -> Vec<f64> {
        let n = u.len();
        let mut result = vec![0.0; n];

        let u_slice = u;

        for (i, r) in result.iter_mut().enumerate() {
            *r = u_slice.iter()
                .enumerate()
                .map(|(j, &u)| self.eval_a(j, i) * u)
                .sum();
        }

        result
    }

    fn eval_ata_times_u(&self, u: &[f64]) -> Vec<f64> {
        let temp = self.eval_a_times_u(u);
        self.eval_at_times_u(&temp)
    }
}

impl Benchmark for Spectralnorm {
    fn name(&self) -> String {
        "Spectralnorm".to_string()
    }

    fn run(&mut self, _iteration_id: i64) {
        self.v = self.eval_ata_times_u(&self.u);
        self.u = self.eval_ata_times_u(&self.v);
    }

    fn checksum(&self) -> u32 {
        let mut vbv = 0.0;
        let mut vv = 0.0;

        for i in 0..(self.size_val as usize) {
            let vi = self.v[i];
            vbv += self.u[i] * vi;
            vv += vi * vi;
        }

        let result_value = (vbv / vv).sqrt();
        helper::checksum_f64(result_value)
    }
}

File: julia/src/benchmarks/Spectralnorm.jl
mutable struct Spectralnorm <: AbstractBenchmark
    size::Int64
    u::Vector{Float64}
    v::Vector{Float64}
    result::UInt32

    function Spectralnorm()
        size_val = Helper.config_i64("Spectralnorm", "size")
        u = fill(1.0, size_val)
        v = fill(1.0, size_val)
        new(size_val, u, v, UInt32(0))
    end
end

name(b::Spectralnorm)::String = "Spectralnorm"

@inline function eval_A(i::Int64, j::Int64)::Float64

    ij = Float64(i + j)
    return 1.0 / ((ij * (ij + 1.0)) / 2.0 + Float64(i) + 1.0)
end

function eval_A_times_u(u::Vector{Float64})::Vector{Float64}
    n = length(u)
    v = Vector{Float64}(undef, n)

    @inbounds for i in 1:n
        s = 0.0
        @simd for j in 1:n
            s += eval_A(i-1, j-1) * u[j]  
        end
        v[i] = s
    end

    return v
end

function eval_At_times_u(u::Vector{Float64})::Vector{Float64}
    n = length(u)
    v = Vector{Float64}(undef, n)

    @inbounds for i in 1:n
        s = 0.0
        @simd for j in 1:n
            s += eval_A(j-1, i-1) * u[j]  
        end
        v[i] = s
    end

    return v
end

function eval_AtA_times_u(u::Vector{Float64})::Vector{Float64}

    Au = eval_A_times_u(u)
    return eval_At_times_u(Au)
end

function run(b::Spectralnorm, iteration_id::Int64)
    b.v = eval_AtA_times_u(b.u)
    b.u = eval_AtA_times_u(b.v)
end

function checksum(b::Spectralnorm)::UInt32
    vBv = vv = 0.0

    @inbounds for i in 1:b.size
        vBv += b.u[i] * b.v[i]
        vv += b.v[i] * b.v[i]
    end

    result = sqrt(vBv / vv)
    return Helper.checksum_f64(result)
end